src/workspaces/common/state/useWorkspaceStatePolling.ts

```diff
@@ -0,0 +1,121 @@
+import { LoadedState } from '@terra-ui-packages/core-utils';
+import _ from 'lodash/fp';
+import { useEffect, useRef, useState } from 'react';
+import { Ajax } from 'src/libs/ajax';
+import { workspacesStore, workspaceStore } from 'src/libs/state';
+import { pollWithCancellation } from 'src/libs/utils';
+import { BaseWorkspaceInfo, WorkspaceState, WorkspaceWrapper as Workspace } from 'src/workspaces/utils';
+
+// type WorkspaceUpdate = Partial<Workspace['workspace']>
+
+interface WorkspaceUpdate extends Partial<BaseWorkspaceInfo> {
+  workspaceId: string;
+  state: WorkspaceState;
+  errorMessage?: string;
+}
+
+const updateWorkspacesList = (workspaces: Workspace[], update: WorkspaceUpdate): Workspace[] =>
+  workspaces.map((ws) => updateWorkspaceIfMatching(ws, update));
+
+const updateWorkspaceIfMatching = <T extends Workspace>(workspace: T, update: WorkspaceUpdate): T => {
+  if (workspace.workspace?.workspaceId === update.workspaceId) {
+    const updatedWs = _.cloneDeep(workspace);
+    updatedWs.workspace.state = update.state;
+    updatedWs.workspace.errorMessage = update.errorMessage;
+    return updatedWs;
+  }
+  return workspace;
+};
+
+const checkWorkspaceState = async (workspace: Workspace, abort: () => void, signal: AbortSignal) => {
```

Comment:
It would be nice to have some documentation to explain what this method does. I'm wondering why an abort function is passed, and what the contract is for when it is called. Also have no idea about how signal is used.

Reply:
Since `abort` seems to be the last thing called, I wonder if this function could return a Promise and then it's caller could deal with `abort` based on whether the Promise is resolved or rejected?

Reply:
I'm giving it a minor restructuring so the returns are a bit cleaner and easier to read.